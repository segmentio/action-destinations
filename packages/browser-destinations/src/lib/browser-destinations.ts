import type { Analytics, Context, Plugin } from '@segment/analytics-next'
import type {
  BaseDefinition,
  BaseActionDefinition,
  ExecuteInput,
  JSONLikeObject,
  InputField
} from '@segment/actions-core'

export type ActionInput<Settings, Payload> = ExecuteInput<Settings, Payload> & {
  analytics: Analytics
  context: Context
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export interface BrowserActionDefinition<Settings, Client, Payload = any> extends BaseActionDefinition {
  /** The operation to perform when this action is triggered */
  perform: (client: Client, data: ActionInput<Settings, Payload>) => Promise<unknown> | unknown

  /**
   * The target platform for the action
   */
  platform: 'web'

  /** Which step in the Analytics.js lifecycle this action should run */
  lifecycleHook?: Plugin['type']
}

export interface BrowserDestinationDependencies {
  loadScript: (src: string, attributes?: Record<string, string>) => Promise<HTMLScriptElement>
  resolveWhen: (condition: () => boolean, timeout?: number) => Promise<void>
}

export type InitializeOptions<Settings> = { settings: Settings; analytics: Analytics }

export interface BrowserDestinationDefinition<Settings = unknown, Client = unknown> extends BaseDefinition {
  mode: 'device'

  /**
   * The function called when the destination has loaded and is ready to be initialized
   * Typically you would configure an SDK or API client here.
   * The return value is injected to your actions as the `client`
   */
  initialize: (options: InitializeOptions<Settings>, dependencies: BrowserDestinationDependencies) => Promise<Client>

  /**
   * Top-level settings that should be available across all actions
   * This is often where you would put initialization settings,
   * SDK keys, API subdomains, etc.
   */
  settings?: Record<string, InputField>

  /** Actions */
  actions: Record<string, BrowserActionDefinition<Settings, Client>>
}

export interface Subscription {
  partnerAction: string
  name: string
  enabled: boolean
  subscribe: string
  mapping: JSONLikeObject
}
