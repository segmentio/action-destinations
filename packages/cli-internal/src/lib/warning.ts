import type { Command } from '@oclif/command'
import chalk from 'chalk'

const warning = chalk.red`
ðŸš¨ðŸš¨ðŸš¨ðŸš¨ðŸš¨ Warning ðŸš¨ðŸš¨ðŸš¨ðŸš¨ðŸš¨ðŸš¨
Internal CLI Commands have been deprecated
please see the action-cli repo for more details
ðŸš¨ðŸš¨ðŸš¨ðŸš¨ðŸš¨ Warning ðŸš¨ðŸš¨ðŸš¨ðŸš¨ðŸš¨ðŸš¨
`

const wait = (ms: number) => {
  return new Promise((resolve) => {
    setTimeout(resolve, ms)
  })
}

const deprecationWarning = async (warn: Command['warn']) => {
  warn(warning)
  await wait(5000)
}

export default deprecationWarning
