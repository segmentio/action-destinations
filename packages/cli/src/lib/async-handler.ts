import { Request, Response, NextFunction, RequestHandler } from 'express'

type AsyncRequestHandler = (req: Request, res: Response, next: NextFunction) => Promise<unknown>

export default function asyncHandler(fn: AsyncRequestHandler): RequestHandler {
  return async (req: Request, res: Response, next: NextFunction): Promise<unknown> => {
    return fn(req, res, next).catch(next)
  }
}
