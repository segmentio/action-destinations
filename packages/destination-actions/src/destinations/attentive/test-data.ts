import { SegmentEvent } from '@segment/actions-core' // Adjust the import as needed for your project

// Define the function with a return type for clarity
export function getCustomEventsTestValidPayload(timestamp: string): Partial<SegmentEvent> {
  return {
    timestamp,
    event: 'Event Type 1',
    messageId: '123e4567-e89b-12d3-a456-426614174000',
    type: 'track',
    userId: '123e4567-e89b-12d3-a456-426614174000',
    context: {
      traits: {
        phone: '+3538675765689',
        email: 'test@test.com'
      }
    },
    properties: {
      tracking_url: 'https://tracking-url.com',
      product_name: 'Product X'
    }
  }
}

export function getCustomEventsTestMapping() {
  return {
    type: { '@path': '$.event' },
    userIdentifiers: {
      phone: { '@path': '$.context.traits.phone' },
      email: { '@path': '$.context.traits.email' },
      clientUserId: { '@path': '$.userId' }
    },
    properties: { '@path': '$.properties' },
    externalEventId: { '@path': '$.messageId' },
    occurredAt: { '@path': '$.timestamp' }
  }
}

export function getCustomEventsTestExpectedPayload(validPayload: Record<string, any>) {
  return {
    type: validPayload.type,
    properties: validPayload.properties,
    externalEventId: validPayload.messageId,
    occurredAt: validPayload.timestamp,
    user: {
      phone: validPayload.context?.traits?.phone,
      email: validPayload.context?.traits?.email,
      externalIdentifiers: {
        clientUserId: validPayload.userId
      }
    }
  }
}

export function getECommEventTestValidPayload(timestamp: string): Partial<SegmentEvent> {
  return {
    timestamp,
    event: 'Product Viewed',
    messageId: '123e4567-e89b-12d3-a456-426614174001',
    type: 'track',
    userId: '123e4567-e89b-12d3-a456-426614174001',
    context: {
      traits: {
        phone: '+3538675765689',
        email: 'test@test.com'
      }
    },
    properties: {
      items: [
        {
          productId: 'prod_123',
          productVariantId: 'var_456',
          productImage: 'https://image-url.com/product.jpg',
          productUrl: 'https://product-url.com',
          name: 'Product X',
          price: {
            value: 29.99,
            currency: 'USD'
          },
          quantity: 1
        }
      ]
    }
  }
}

export function getECommEventTestMapping() {
  return {
    items: [
      {
        productId: { '@path': '$.properties.items[0].productId' },
        productVariantId: { '@path': '$.properties.items[0].productVariantId' },
        productImage: { '@path': '$.properties.items[0].productImage' },
        productUrl: { '@path': '$.properties.items[0].productUrl' },
        name: { '@path': '$.properties.items[0].name' },
        price: {
          value: { '@path': '$.properties.items[0].price.value' },
          currency: { '@path': '$.properties.items[0].price.currency' }
        },
        quantity: { '@path': '$.properties.items[0].quantity' }
      }
    ],
    userIdentifiers: {
      phone: { '@path': '$.context.traits.phone' },
      email: { '@path': '$.context.traits.email' },
      clientUserId: { '@path': '$.userId' }
    },
    occurredAt: { '@path': '$.timestamp' }
  }
}

export function getECommEventTestExpectedPayload(validPayload: Record<string, any>) {
  return {
    items: validPayload.properties?.items,
    occurredAt: validPayload.timestamp,
    user: {
      phone: validPayload.context?.traits?.phone,
      email: validPayload.context?.traits?.email,
      externalIdentifiers: {
        clientUserId: validPayload.userId
      }
    }
  }
}
