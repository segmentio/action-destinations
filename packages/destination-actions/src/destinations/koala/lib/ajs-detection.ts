export interface AnalyticsEventContext {
  library?: {
    name?: string
  }
}

export function isAJSEvent(eventContext?: AnalyticsEventContext): boolean {
  return eventContext?.library?.name === 'analytics.js'
}
