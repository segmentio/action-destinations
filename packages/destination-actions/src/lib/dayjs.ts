import dayjs from 'dayjs'
import utc from 'dayjs/plugin/utc'
import advancedFormat from 'dayjs/plugin/advancedFormat'

dayjs.extend(utc)
dayjs.extend(advancedFormat)

export default dayjs
