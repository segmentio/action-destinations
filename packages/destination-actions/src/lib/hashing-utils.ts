/**
 * Utility module for hashing values using various encryption methods and digest types.
 * It includes functionality to check if a value is already hashed and to process hashing with optional cleaning.
 */
import * as crypto from 'crypto'
import { Features } from '@segment/actions-core'

export const EncryptionMethods = ['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512'] as const
export type EncryptionMethod = typeof EncryptionMethods[number]

export const DigestTypes = ['hex', 'base64'] as const
export type DigestType = typeof DigestTypes[number]

type CleaningFunction = (value: string) => string

export const hashConfigs: {
  [key in EncryptionMethod]: { lengthHex: number; lengthBase64: number }
} = {
  md5: { lengthHex: 32, lengthBase64: 24 },
  sha1: { lengthHex: 40, lengthBase64: 28 },
  sha224: { lengthHex: 56, lengthBase64: 40 },
  sha256: { lengthHex: 64, lengthBase64: 44 },
  sha384: { lengthHex: 96, lengthBase64: 64 },
  sha512: { lengthHex: 128, lengthBase64: 88 }
}

const slugsToBypassFeatureFlag = ['actions-facebook-custom-audiences']

class SmartHashing {
  private preHashed: boolean

  /**
   * Creates an instance of SmartHashing.
   * @param encryptionMethod - The method of encryption to be used.
   * @param digest - The type of digest to be used.
   */
  constructor(public encryptionMethod: EncryptionMethod = 'sha256', public digest: DigestType = 'hex') {
    this.preHashed = false
  }

  isAlreadyHashed(value: string): boolean {
    const config = hashConfigs[this.encryptionMethod]
    if (!config) throw new Error(`Unsupported encryption method: ${this.encryptionMethod}`)

    let regex: RegExp
    switch (this.digest) {
      case 'hex':
        regex = new RegExp(`^[a-f0-9]{${config.lengthHex}}$`, 'i')
        this.preHashed = value.length === config.lengthHex && regex.test(value)
        break
      case 'base64':
        regex = new RegExp(`^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$`)
        this.preHashed = value.length === config.lengthBase64 && regex.test(value)
        break
      default:
        throw new Error(`Unsupported digest type: ${this.digest}`)
    }

    return this.preHashed
  }

  hash(value: string): string {
    if (value.trim() === '') {
      throw new Error('Cannot hash an empty string')
    }

    if (this.preHashed) {
      return value
    }

    return crypto.createHash(this.encryptionMethod).update(value).digest(this.digest)
  }
}

/**
 * Processes the hashing of a given value based on the provided encryption method, digest type, features, and optional cleaning function.
 *
 * @param value - The string value to be hashed.
 * @param encryptionMethod - The method of encryption to be used.
 * @param digest - The type of digest to be used.
 * @param features - An object containing feature flags.
 * @param destinationSlugForBypass - A slug that, if present in the bypassFlagSlugs array, will bypass the flag check.
 * @param cleaningFunction - An optional function to clean the value before hashing.
 * @returns The hashed value or the original value if it is already hashed.
 */
export function processHashing(
  value: string,
  encryptionMethod: EncryptionMethod,
  digest: DigestType,
  features: Features,
  destinationSlugForBypass: string,
  cleaningFunction?: CleaningFunction
): string {
  if (value.trim() === '') {
    return ''
  }

  const smartHashing = new SmartHashing(encryptionMethod, digest)
  /**
   * Determines whether the flag should be bypasssed or not.
   * If the slug is present in the bypassFlagSlugs array, the flag check will be bypassed.
   */
  const bypassFlag = destinationSlugForBypass && slugsToBypassFeatureFlag.includes(destinationSlugForBypass)

  // If smart-hashing feature flag is not enabled, clean and hash the value directly
  if (!bypassFlag && !(features && features['smart-hashing'])) {
    if (cleaningFunction) {
      value = cleaningFunction(value)
    }
    return smartHashing.hash(value)
  }

  if (smartHashing.isAlreadyHashed(value)) {
    return value
  }

  if (cleaningFunction) {
    value = cleaningFunction(value)
  }

  return smartHashing.hash(value)
}
