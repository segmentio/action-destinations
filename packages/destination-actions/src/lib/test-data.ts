/* eslint-disable @typescript-eslint/no-explicit-any */
import { ActionDefinition, DestinationDefinition } from '@segment/actions-core'
import Chance from 'chance'

function setTestData(seedName: string, type: string, fieldName?: string, format?: string) {
  const chance = new Chance(seedName)

  let val: any
  switch (type) {
    case 'boolean':
      val = chance.bool()
      break
    case 'datetime':
      val = chance.date()
      break
    case 'integer':
      val = chance.integer()
      break
    case 'number':
      val = chance.floating({ fixed: 2 })
      break
    case 'text':
      val = chance.sentence()
      break
    case 'object':
      val = { testType: chance.string() }
      break
    default:
      val = chance.string()
      break
  }

  if (format === 'uri') val = chance.url()
  if (fieldName === 'email') val = chance.email()
  if (fieldName === 'currency') val = chance.currency().code

  return val
}

function setData(eventData: any, chanceName: string, fieldName: string, field: any, data?: any) {
  const { format, multiple, type } = field

  if (!data) {
    data = setTestData(chanceName, type, fieldName, format)
  }

  eventData[fieldName] = multiple ? [data] : data
  return eventData
}

export function generateTestData(
  seedName: string,
  destination: DestinationDefinition<any>,
  action: ActionDefinition<any>,
  isRequiredOnly: boolean
) {
  let eventData: any = {}
  const settingsData: any = {}

  const authentication = destination.authentication
  if (authentication) {
    for (const settingKey in authentication.fields) {
      const { format, type } = authentication.fields[settingKey]
      settingsData[settingKey] = setTestData(seedName, type, undefined, format)
    }
  }

  for (const [name, field] of Object.entries(action.fields)) {
    if (isRequiredOnly && !(field.required || name.includes('id'))) {
      continue
    }

    const { properties } = field

    if (properties) {
      let subData: any = {}
      let propertyFields = Object.keys(properties)

      if (isRequiredOnly) {
        propertyFields = propertyFields.filter((name) => properties[name].required)
      }

      for (const propertyName of propertyFields) {
        const property = properties[propertyName]
        subData = setData(subData, seedName, propertyName, property)
      }

      eventData = setData(eventData, seedName, name, field, subData)
      continue
    }

    eventData = setData(eventData, seedName, name, field)
  }

  return [eventData, settingsData]
}
