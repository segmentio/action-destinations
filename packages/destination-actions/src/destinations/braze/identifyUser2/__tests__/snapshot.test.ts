import { createTestEvent, createTestIntegration } from '@segment/actions-core'
import { generateTestData } from '../../../../lib/test-data'
import destination from '../../index'
import nock from 'nock'

const testDestination = createTestIntegration(destination)
const actionSlug = 'identifyUser2'
const destinationSlug = 'Braze'
const seedName = `${destinationSlug}#${actionSlug}`

describe(`Testing snapshot for ${destinationSlug}'s ${actionSlug} destination V2 action:`, () => {
  it('fails if sync mode is delete', async () => {
    const action = destination.actions[actionSlug]
    const [eventData, settingsData] = generateTestData(seedName, destination, action, false)

    const event = createTestEvent({
      properties: eventData
    })

    await expect(
      testDestination.testAction(actionSlug, {
        event: event,
        mapping: { ...event.properties, __segment_internal_sync_mode: 'delete' },
        settings: settingsData,
        auth: undefined
      })
    ).rejects.toThrowError('syncMode must be "add" or "upsert"')
  })

  it('fails if sync mode is update', async () => {
    const action = destination.actions[actionSlug]
    const [eventData, settingsData] = generateTestData(seedName, destination, action, false)

    const event = createTestEvent({
      properties: eventData
    })

    await expect(
      testDestination.testAction(actionSlug, {
        event: event,
        mapping: { ...event.properties, __segment_internal_sync_mode: 'update' },
        settings: settingsData,
        auth: undefined
      })
    ).rejects.toThrowError('syncMode must be "add" or "upsert"')
  })

  it('snapshot with all fields', async () => {
    const action = destination.actions[actionSlug]
    const [eventData, settingsData] = generateTestData(seedName, destination, action, false)

    nock(/.*/).persist().get(/.*/).reply(200)
    nock(/.*/).persist().post(/.*/).reply(200)
    nock(/.*/).persist().put(/.*/).reply(200)

    const event = createTestEvent({
      properties: eventData
    })

    const responses = await testDestination.testAction(actionSlug, {
      event: event,
      mapping: { ...event.properties, __segment_internal_sync_mode: 'add' },
      settings: settingsData,
      auth: undefined
    })

    const request = responses[0].request
    const rawBody = await request.text()

    try {
      const json = JSON.parse(rawBody)
      expect(json).toMatchSnapshot()
      return
    } catch (err) {
      expect(rawBody).toMatchSnapshot()
    }
  })
})
