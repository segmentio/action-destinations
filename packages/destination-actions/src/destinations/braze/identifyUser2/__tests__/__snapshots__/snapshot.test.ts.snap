// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing snapshot for Braze's identifyUser2 destination V2 action: snapshot with all fields 1`] = `
Object {
  "aliases_to_identify": Array [
    Object {
      "external_id": "O$J@g9^nDz",
      "user_alias": Object {
        "alias_label": "O$J@g9^nDz",
        "alias_name": "O$J@g9^nDz",
      },
    },
  ],
  "merge_behavior": "none",
}
`;
